### PR TITLE
Make spirv-cross optional

### PR DESCRIPTION
**Pull request by kvark**
SPIRV-Cross is a big foreign dependency, which often cause issues, such as https://github.com/denoland/rusty_v8/issues/465 and #3389, #3117, #2520, and many others.
Previously in #3445 and #3520 we implemented Naga code path as an alternative, but hidden behind an optional feature. Now it's time to flip the table and say that Naga is permanent, but SPIRV-Cross is optional until completely phased out.

Most of the changes here are from re-routing the main shader creation logic from using SPIRV-Cross types to using Naga types. The "cross" stuff is just getting derived from Naga types if necessary.

The change only applies to Metal and OpenGL backends so far. D3D will come on board once https://github.com/gfx-rs/naga/pull/408 is operational.

Note: it's going to be a long while for Naga to mature, most users need to keep the "cross" dependency enabled for now! Even the included examples don't run on Naga yet, since the SPIR-V generated by glslang is quite specific, and Naga's SPV frontend doesn't like it yet.

### CODE REVIEW

File: src/backend/gl/src/device.rs [hunk 1]
```diff
@@ -519,133 +513,137 @@ impl Device {
...
+            self.combine_separate_images_and_samplers(&mut ast, context.reborrow());
+            self.set_push_const_layout(&mut ast).unwrap();
+
+            let glsl = self
+                .translate_spirv_cross(&mut ast, stage, ep.entry)
+                .unwrap();
+            debug!("SPIRV-Cross generated shader:\n{}", glsl);
+            result = Self::create_shader_module_raw(&self.share.context, &glsl, stage);
+        }
+        if !ep.module.prefer_naga {
```
Comment (Gordon-F) [inline src/backend/gl/src/device.rs:636]: @kvark Awesome work! Thank you!

Maybe you mean:
```rust
if ep.module.prefer_naga {
```

Otherwise, we still get naga errors with the enabled `cross` feature.

Comment (kvark) [reply src/backend/gl/src/device.rs]: The point here is to try Naga in case it was not preferred originally. I was indeed missing a check for the result here, thank you for spotting!

**Review by grovesNL — APPROVED** [review]
Thanks! It was a bit hard to follow the MSL changes in detail because of things moving around, but overall everything seems alright with me

File: src/backend/metal/Cargo.toml [hunk 2]
```diff
@@ -16,36 +16,45 @@ edition = "2018"
 default = []
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/backend/metal/Cargo.toml:16]: That's an interesting idea! I'm generally opposed to default features. If you need it, it's a one time change to add them. If you don't, it's a whole mess out of a sudden.

Given that we are changing gfx-rs on git here, and the issue will likely be discovered instantly if anybody faces it, it seems to me that the risk to cause trouble with this change is very low.

**Comment by kvark** [comment]
Thank you for the review!
Let's keep the option to enable it by default on the table. I'm still thinking about this.
bors r=grovesNL
